drivers/ti/uart/aarch64/16550_console.S

```diff
@@ -137,6 +137,9 @@ func console_16550_core_putc
 #if ENABLE_ASSERTIONS
 	cmp	x1, #0
 	ASM_ASSERT(ne)
+#else
```

Comment:
Always exit when the base is zero is fine with me. I think it was the older code behavior (maybe 1.2 or 1.3). 
Which information could be provided by ASM_ASSERT if the UART base is zero? I would expect some error message on the console, which is in this case not initialized.
Would it be acceptable removing "#if ENABLE_ASSERTION" block and using the code from this patch (check if zero and return if yes)?

Reply:
>  Which information could be provided by ASM_ASSERT if the UART base is zero? I would expect some error message on the console, which is in this case not initialized.

Heh, that's a good point. However, I think that the assert is still useful. You may be trying to print to an uninitialized console, and assert will print to the plat crash console. The crash console may or may not use the multi console API. Arm platforms, RPi3, QEMU, Allwinner... they all use direct calls to the console core functions instead of going through the multi console API, so you can be sure to get the error message.

Maybe what we need is a dummy console driver.

Reply:
Maybe what you need to use is the multi console driver. With it you can register multiple/one/no console drivers, and you could get the result you want. Basically, you can run with no console registered (and you can even register a runtime console when you are about to leave BL31, for example). You could also not register the console in release builds, if that's what you want.

If you force push to the PR, the old patches are replaced by the new ones. It makes it hard to compare versions of the same PR, but some comments stay in place, so it's better than nothing.

Reply:
I definitely planned to switch to the new console API. It was not working when I tried it first (stop running somewhere before BL3 load). Since I was busy with the other stuff, I decided to debug this issue and switch to the new API bit later. Currently my target is to create an initial code base for Marvell platforms in TF-A (no trusted boot, single platform only, etc.) and then push additional features and platforms one by one.
For archiving that I can drop some patches if a quick fix is not an option. Specifically, if I drop this patch, I will loose an option to download recovery image over serial line. This is very useful feature allowing to de-brick system with bad boot image without JTAG, but it can be disabled for the first patch series.

Reply:
I agree that it's better not to modify more things for the first version.

Then I think it's better if you drop the patch from the upstream port as there is no code to load images through the serial port so the upstream code doesn't benefit from this. After it is merged and you migrate to the multi console driver you won't need this patch.

Reply:
The recovery image load is a BootROM feature, so there is no special boot image required. A regular flash image created by TF-A build process can be downloaded over the board serial line for bringing the system up after a bad update.
The helper script for doing so is included in u-boot "tools" folder - https://github.com/u-boot/u-boot/blob/master/tools/mrvl_uart.sh
So this feature is quite useful if included.